src/boot/StoreHydratedGate.js

```diff
@@ -0,0 +1,57 @@
+/* @flow strict-local */
+import React, { useCallback } from 'react';
+import type { Node } from 'react';
+
+import { useEdgeTriggeredEffect } from '../reactUtils';
+import { useGlobalSelector, useDispatch } from '../react-redux';
+import { getIsHydrated, getHasAuth } from '../selectors';
+import FullScreenLoading from '../common/FullScreenLoading';
+import { registerAndStartPolling } from '../events/eventActions';
+import { sendOutbox } from '../outbox/outboxActions';
+import { initNotifications } from '../notification/notifTokens';
+
+type Props = $ReadOnly<{|
+  children: Node,
+|}>;
+
+/**
+ * Where we prevent everything from rendering while waiting for rehydration.
+ *
+ * Also where we dispatch post-rehydrate actions.
+ */
+export default function StoreHydratedGate(props: Props): Node {
+  const { children } = props;
+
+  const dispatch = useDispatch();
+
+  const isHydrated = useGlobalSelector(getIsHydrated);
+  const hasAuth = useGlobalSelector(getHasAuth);
+
+  const init = useCallback(async () => {
+    // Init right away if there's an active, logged-in account.
+    // NB `getInitialRouteInfo` depends intimately on this behavior.
+    if (hasAuth) {
+      await dispatch(registerAndStartPolling());
+
+      // TODO(#3881): Lots of issues with outbox sending
+      dispatch(sendOutbox());
+
+      dispatch(initNotifications());
+    }
+  }, [dispatch, hasAuth]);
+
+  useEdgeTriggeredEffect(
+    () => {
+      init();
+    },
```

Comment:
`init` returns a Promise.

`useEdgeTriggeredEffect`, as well as `useEffect` and our new `useConditionalEffect` we discussed in the office on Thursday, won't take a callback that returns a Promise.

In fact, if I just pass an async function expression directly to `useEffect`—

```js
useEffect(async () => 'foo');
```

—I get a complaint from `react-hooks/exhaustive-deps`:

```
Effect callbacks are synchronous to prevent race conditions. Put the async function inside:

useEffect(() => {
  async function fetchData() {
    // You can await here
    const response = await MyAPI.getData(someId);
    // ...
  }
  fetchData();
}, [someId]); // Or [] if effect doesn't need props or state

Learn more about data fetching with Hooks: https://reactjs.org/link/hooks-data-fetchingeslint[react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920)
```

Reply:
I see, yeah. Here we're happy for it to be an async thing we don't wait for, but it is good to have something making that explicit. Which this does, a little bit.